### ISSUE TITLE
[research] AERA-v26.10 latent-depth sync coalescing after #702/#705

### ISSUE THREAD

**Issue by vinceackermann2-sys** [vinceackermann2-sys/tam-research#706]
## Purpose

Implement exactly the concrete `launch_or_idle_reduction` mechanism selected by authoritative #702/#705, without changing the learned architecture: coalesce the hard latent-reasoner’s four per-step CUDA `nonzero()`/dynamic-compaction synchronization points into one stable depth-prefix schedule per stage invocation.

## Exact source and parent evidence

Bind current source main/tree at preregistration:
- `SOURCE_MAIN=4883abe053347379812901c5c2057da70c11e3e7`
- `SOURCE_TREE=27be25ba578aa530c1e6d832db1a38be4c8bc22a`

Authoritative parent diagnostic:
- parent issue `#702`
- sole L4 trigger `#705`
- evidence comment `5572472746`
- run `34135719600`
- job `101786085046`
- attempt `1`
- one NVIDIA L4
- source systems decision remains `FAIL_FROZEN_E2E_SYSTEMS_GATE`
- diagnostic `next_target=launch_or_idle_reduction`
- optimization_authorized by parent result: `false`

Observed full-call CUDA runtime evidence from #705:
- batch 8: 1564 CUDA kernel launches; 50 synchronize/wait runtime calls = 49 `cudaStreamSynchronize` + 1 `cudaDeviceSynchronize`; parent unresolved lower-level gap `9.2633691290416 ms`
- batch 64: 1594 CUDA kernel launches; 52 synchronize/wait runtime calls = 51 `cudaStreamSynchronize` + 1 `cudaDeviceSynchronize`; parent unresolved lower-level gap `12.370303117411762 ms`
- kernel↔ATen External-id attribution was incomplete, so no specific-kernel optimization is authorized.

The #705 trace-derived idle numbers are classification evidence only and MUST NOT be treated as directly removable parent-gap milliseconds.

## Source-code localization

The current hard-sparse runtime contains three variable-population `nonzero()` families:
1. optional-stage routing in `HardwareAwareAERATextLMV26._route_one_stage`;
2. second-expert selection in `ExecutionEquivalentNoHostBMMHardSparseExpertBank._hard_forward`;
3. latent-depth selection once per recurrent step in `ExecutionEquivalentNoHostDtypeSafeChunkLatentReasoner.forward`.

The latent reasoner has frozen `max_reason_steps=4`. Therefore each invoked stage may perform up to four separate `nonzero()` compactions before `index_select -> GRUCell -> index_copy`. With the #705 two-chunk/four-stage route geometry this family alone can account for up to 32 variable-output compactions per full call. This successor changes only that repeated latent-depth family. Optional-stage and expert selection remain byte/semantic-equivalent to current v26.9.

Historical #258 already rejected `torch.compile alone` as the systems solution (~1.09-1.12x isolated sparse-reasoner speedup) and directed future work toward semantics-preserving coalesced/fused execution. This successor is not a retry of #258.

## Frozen implementation mechanism

For hard latent reasoning only:
- preserve `chosen = depth_logits.argmax(dim=-1) + 1` exactly;
- build one stable descending depth permutation on device, preserving original row order within equal chosen depth;
- compute all four active-prefix sizes from the same `chosen` tensor and materialize those four small counts to host in one explicit synchronization;
- gather the current latent state once into depth-prefix order;
- for each recurrent step, execute the exact inherited GRUCell only on the active prefix for that step;
- update only that active prefix;
- restore original row order once after the final step;
- preserve device-resident detached telemetry (`last_steps`, `last_expected`) and all dtype/autocast behavior.

The number of GRUCell updates per example must remain exactly its frozen chosen depth. Inactive rows may not execute GRUCell work. Dense masking receives zero sparse credit.

No stage-routing decision, expert decision, depth decision, controller equation, attention, FICEM equation/backend, memory write, durable state, checkpoint byte, parameter, parameter schema, state-dict schema, token, batch, scientific seed, or training behavior may change.

Prefer an additive v26.10 execution wrapper/installer rather than mutating the frozen v26.9 source files.

## CPU-first implementation governance

Before any branch/ref move:
1. create issue-specific implementation/launcher/workflow/test candidate blobs as detached Git objects;
2. freeze all exact dependency blobs, result path, trigger prefixes, thresholds, source tree, new blobs, detached tree, and detached commit in one comment;
3. only then create/move the named research branch;
4. open one immutable PR and consume exactly one original full-repository CPU CI attempt;
5. never rerun/retry/redispatch a consumed attempt.

No Modal/GPU execution is authorized by this issue or by CPU CI.

## Future one-L4 optimization microbenchmark gate

Only after CPU-green merge, fresh exact-main audit, successful zero-GPU preauthorization, and a separate exact-main L4 authorization may one fresh L4 optimization trigger execute.

Frozen fixtures:
- checkpoint `/vol/aera-real-language/v25-dev-seed8471`
- batches `(8,64)`
- token seed rule `138471 + 10000 + batch_size`
- `route_mode="hard_sparse"`
- `hard=True`
- `update_memory=True`
- CUDA BF16 autocast
- compare current merged v26.9 baseline versus additive v26.10 candidate loaded from identical checkpoint bytes.

Correctness/semantic requirements before any performance credit:
- exact same route gates at every chunk/stage;
- exact same chosen expert counts/IDs;
- exact same chosen latent depths;
- exact same per-step active-row counts;
- same parameter/state-dict schema and checkpoint hashes;
- finite outputs;
- logits and continuous durable state satisfy the inherited frozen integrated tolerances `atol=1e-2`, `rtol=1e-2`;
- boolean validity/state geometry exact;
- no dense-masked execution;
- no training/optimizer/backward/checkpoint mutation.

Timing/mechanism gate, both batches required:
- separately measured unprofiled CUDA-event full-call medians after frozen warmup;
- candidate full-call median latency must improve by at least `5%` versus current v26.9 baseline at batch 8 AND batch 64;
- one low-intrusion post-timing profiler call may count runtime synchronizations; candidate must show at least `20` fewer `cudaStreamSynchronize` calls inside the full-call boundary than baseline at batch 8 AND batch 64;
- profiler-distorted timings cannot substitute for the unprofiled latency gate.

If any correctness/sparsity gate fails, or either batch misses either performance threshold, decision is FAIL and no full E2E systems rerun is authorized.

If all gates pass, this authorizes only a separately preregistered fresh full frozen Transformer-relative E2E systems gate. It does NOT itself earn systems PASS.

## Authority ceiling

No architecture freeze, S2, fresh scientific seed, independent replication credit, 100M scaling, or breakthrough claim is authorized here.

**Comment by vinceackermann2-sys** [comment]
## #706 pre-implementation v26.10 latent-depth sync coalescing freeze

This freezes the additive execution-only candidate before any branch/ref move. No GPU/Modal execution, optimization result, systems PASS, architecture freeze, S2, fresh scientific seed, replication, scaling, or breakthrough authority is created by this comment.

SOURCE_MAIN=4883abe053347379812901c5c2057da70c11e3e7
SOURCE_TREE=27be25ba578aa530c1e6d832db1a38be4c8bc22a
BRANCH=research/aera-v26-10-issue706-latent-depth-sync-coalescing
PREAUTH_PREFIX=[aera-v26-10-issue706-latent-depth-sync-coalescing-preauth]
L4_PREFIX=[aera-v26-10-issue706-latent-depth-sync-coalescing-l4]
RESULT_PATH=/vol/aera-v26/issue706-v26-10-latent-depth-sync-coalescing/result.json
PARENT_RESULT_PATH=/vol/aera-v26/issue702-cuda-timeline-dispatch-gap-attribution/result.json
PARENT_ISSUE=702
PARENT_TRIGGER=705
PARENT_EVIDENCE_COMMENT=5572472746
PARENT_RUN=34135719600
PARENT_JOB=101786085046
PARENT_ATTEMPT=1
SOURCE_DECISION=FAIL_FROZEN_E2E_SYSTEMS_GATE
PARENT_NEXT_TARGET=launch_or_idle_reduction

CHECKPOINT_PATH=/vol/aera-real-language/v25-dev-seed8471
CHECKPOINT_AERA_SHA256=f8aa92421801e8f190247e420632be5f0c20bc5ea8bf6bdeefe06686b3a31b30
CHECKPOINT_TRANSFORMER_SHA256=cdd5cab4439a709468d6607d45d82081b33e876b2e40d91d4a38ba139b219dd7

V26_10_IMPL_BLOB=d8f691c198eed1fa96bcbb78a4e76cad82d18779
LAUNCHER_BLOB=e23072aa83b0bc20ccd72e7567f234e9f4a931f6
WORKFLOW_BLOB=c842f636797d7a5c424d2dc8718f023efc2efe90
CPU_TEST_BLOB=99dd0f6a6fbae273326004def8afd23cf7c681ca
V25_1_NOHOST_BLOB=237e5615cf32f644e8675808a6b3e9adaf04fb23
V26_RUNTIME_BLOB=268644ac4edee15a4cc4e29d3fed7f61eeb3caa7
V26_9_SYSTEMS_BLOB=512572340cc09e2e7ad6729712258c12cb377ef2
BASE_SYSTEMS_BLOB=c9731cae7e386f09b2a190b045532591c4fa00be
TRIAGE_BLOB=e5ce8cdda0777dce97816e2640f4492803a6b191
V26_9_BACKEND_BLOB=b81cc209f5d95abbe1fb8bd620c78e87c067bc19
ISSUE665_LAUNCHER_BLOB=72f27391ff2f0a7bff8d4532f307ddc4869cf494

BATCHES=8,64
TOKEN_SEED_RULE=138471+10000+batch_size
ROUTE_MODE=hard_sparse
HARD=True
UPDATE_MEMORY=True
CUDA_AUTOCAST=BF16
WARMUP_CALLS=3
TIMED_CALLS_PER_CONDITION=20
PROFILE_CALLS_PER_CONDITION=1
MIN_LATENCY_IMPROVEMENT=0.05
MIN_STREAM_SYNCHRONIZE_REDUCTION=20
INTEGRATED_ATOL=0.01
INTEGRATED_RTOL=0.01

FROZEN_TREE=ff3913de8e605aba10f87839676c9b582304233c
FROZEN_COMMIT=5cac6325fb3047f4a573d5b6e8eeeda202834c16

Frozen mechanism: preserve `chosen = depth_logits.argmax(dim=-1) + 1`; one stable descending depth permutation; one four-count host materialization per reasoner invocation; exact inherited GRUCell only on active prefixes; one restoration to original row order; unchanged optional-stage/expert/FICEM/write/checkpoint/scientific semantics. Dense-masked execution receives zero sparse credit.

CPU-first continuation only: create the named branch at exactly `FROZEN_COMMIT`, open one immutable PR, and consume exactly one original full-repository CPU CI attempt. Never rerun/retry/redispatch. Only a CPU-green merged exact head plus fresh exact-main audit, one successful zero-GPU preauth, and a separate exact-main L4 authorization can permit one L4 microbenchmark trigger.

**Comment by vinceackermann2-sys** [comment]
## #706 sole L4 latent-depth sync coalescing authorization

Authorize main: `c2e9c17d7bb34d7652fc6759c80193f2d951f2ce`

Authorize exactly one fresh L4 optimization microbenchmark trigger under frozen prefix `[aera-v26-10-issue706-latent-depth-sync-coalescing-l4]`, and nothing else.

Successful required preauthorization:
- trigger issue `#708`
- evidence comment `5573116950`
- workflow run `34140899443`
- job `101802424004`
- attempt `1`
- `result_absent=true`
- `gpu_used=false`
- `model_constructed=false`
- `new_measurement_performed=false`

Frozen implementation boundary:
- source merge main `c2e9c17d7bb34d7652fc6759c80193f2d951f2ce`
- merged tree `ff3913de8e605aba10f87839676c9b582304233c`
- frozen candidate commit `5cac6325fb3047f4a573d5b6e8eeeda202834c16`
- v26.10 implementation blob `d8f691c198eed1fa96bcbb78a4e76cad82d18779`
- launcher blob `e23072aa83b0bc20ccd72e7567f234e9f4a931f6`
- workflow blob `c842f636797d7a5c424d2dc8718f023efc2efe90`
- CPU test blob `99dd0f6a6fbae273326004def8afd23cf7c681ca`
- checkpoint `/vol/aera-real-language/v25-dev-seed8471` with frozen hashes
- batches `(8,64)`, token seed rule `138471 + 10000 + batch_size`, hard-sparse routing, `hard=True`, `update_memory=True`, CUDA BF16 autocast
- latency gate `>=5%` improvement at both batches using separately measured unprofiled CUDA-event medians
- mechanism gate `>=20` fewer `cudaStreamSynchronize` calls at both batches
- all preregistered correctness/sparsity/state/schema/checkpoint gates remain mandatory before performance credit.

This authorization permits only the one #706 L4 optimization microbenchmark. It does not authorize a full E2E systems rerun and cannot itself earn systems PASS. No architecture freeze, S2, fresh scientific seed, independent replication credit, 100M scaling, or breakthrough claim is authorized. The L4 trigger and its Actions attempt are consumed regardless of outcome; never rerun/retry/redispatch.
